schema_parse.go

```diff
@@ -594,3 +642,24 @@ func (c seenCache) Add(name string) error {
 	c[name] = struct{}{}
 	return nil
 }
+
+func preserveLogicalType(logicalType string, props *map[string]any) {
+	if logicalType == "" {
+		return // nothing to preserve
+	}
+	if *props == nil {
+		*props = make(map[string]any, 1)
+	}
+	(*props)["logicalType"] = logicalType
+}
+
+func checkLogicalType(props map[string]any) error {
+	val, ok := props["logicalType"]
+	if !ok {
+		return nil
+	}
+	if _, isString := val.(string); !isString {
+		return fmt.Errorf(`"logicalType" attribute must be a string, got %T`, val)
+	}
+	return nil
+}
```

Comment:
Out of interest, what necessitates `logicalType` in this context be a string? At this point it is metadata, and by spec could be any type.

Reply:
This is for consistency. The way `logicalType` is parsed for other fixed and primitive types, it will be an error if it is not a string. Why would it be an error there but _not_ for the other kinds of types? The [spec](https://avro.apache.org/docs/1.12.0/specification/#logical-types) doesn't say anything about the attribute being specific to fixed or primitive types:

>A logical type is an Avro primitive or complex type with extra attributes to represent a derived type. The attribute `logicalType` must always be present for a logical type, and is a string with the name of one of the logical types listed later in this section. Other attributes may be defined for particular logical types.

So if it's _not_ an error here, and we allow treating a non-string value as an extra property, then I should probably also change the way fixed and primitive types are parsed, to similarly allow non-string values which would result in it just being another property.

Reply:
Okay, I reworked it so that logical type does not have to be a string; it's just another property. For fixed and primitive types, if it's not a string, it's also tossed into the set of other properties and otherwise ignored.